activerecord/lib/active_record/associations/belongs_to_association.rb

```diff
@@ -29,7 +29,7 @@ def handle_dependency
       end
 
       def inversed_from(record)
-        replace_keys(record)
+        replace_keys(record) if owner[reflection.foreign_key] != record_target_key(record)
```

Comment:
This will now read the attribute twice. Since `replace_key` is internal we should pass the key as second argument to avoid the extra lookup.

Reply:
I think you mean the call to `record_target_key` right? I changed it so it is only called once, albeit I'm not sure if there is a better default for the second parameter of `replace_keys`.